src/CodeGen_Hexagon.cpp

```diff
@@ -102,30 +103,183 @@ bool is_dense_ramp(const Expr &x) {
+                if (a.first.defined() && b.first.defined()) {
+                    return {a.first < b.first, a.second < b.second};
+                }
+            }
+        } else if (const LE *op = e.as<LE>()) {
+            if (!op->a.type().is_bool()) {
+                auto a = get_extreme_lanes(op->a);
+                auto b = get_extreme_lanes(op->b);
+                if (a.first.defined() && b.first.defined()) {
+                    return {a.first <= b.first, a.second <= b.second};
+                }
+            }
+        } else if (const Variable *op = e.as<Variable>()) {
+            if (monotonic_vectors.contains(op->name)) {
+                return monotonic_vectors.get(op->name);
+            }
+        } else if (const Let *op = e.as<Let>()) {
+            auto v = get_extreme_lanes(op->value);
+            ScopedBinding<std::pair<Expr, Expr>> bind(v.first.defined(), monotonic_vectors, op->name, v);
+            return get_extreme_lanes(op->body);
+        }
+        return {Expr(), Expr()};
+    }
+
+    Expr visit(const Let *op) override {
+        auto v = get_extreme_lanes(op->value);
+        ScopedBinding<std::pair<Expr, Expr>> bind(op->value.type().is_vector() && v.first.defined(),
+                                                  monotonic_vectors, op->name, v);
+        return IRMutator::visit(op);
+    }
+
     Expr visit(const Load *op) override {
-        // Don't handle loads with without predicates, scalar predicates, or
-        // non-dense ramps.
-        if (is_one(op->predicate) || op->predicate.as<Broadcast>() ||
-            !is_dense_ramp(op->index)) {
+        if (is_one(op->predicate)) {
+            // These are handled fine
             return IRMutator::visit(op);
         }
 
         Expr predicate = mutate(op->predicate);
         Expr index = mutate(op->index);
 
-        // Make the predicate into a scalar that is true if any of the lanes are
-        // true.
-        Expr condition = Shuffle::make({predicate}, {0});
-        for (int i = 1; i < op->type.lanes(); i++) {
-            condition = condition || Shuffle::make({predicate}, {i});
-        }
-        predicate = Broadcast::make(condition, predicate.type().lanes());
+        if (is_dense_ramp(index)) {
+            // Make the predicate into a scalar that is true if any of the lanes are
+            // true.
+
+            Expr condition;
 
-        return Load::make(op->type, op->name, index, op->image, op->param,
-                          predicate, op->alignment);
+            // If the predicate is monotonic increasing or decreasing
+            // over the vector lanes, we can just check the last or
+            // first lane, respectively. We won't bother to
+            // distinguish between the two cases though, so we just or
+            // them both together.
+            auto v = get_extreme_lanes(predicate);
+            if (v.first.defined()) {
+                condition = simplify(v.first || v.second);
```

Comment:
Consider internal_assert that v.first and v.second are bools? I think it would also help a little with readability of this algorithm.